controller/engine_image_controller.go

```diff
@@ -468,28 +463,25 @@ func limitAutomaticEngineUpgradePerNode(volumes map[string]*longhorn.Volume, upg
 //   1. Volume is not in engine upgrading process
 //   2. newEngineImageResource is deployed on attaching node and the all volume's replicas' nodes
 //   3. Volume is in attached state and it is able to do live upgrade
-func (ic *EngineImageController) getVolumesForEngineImageUpgrading(volumes map[string]*longhorn.Volume, newEngineImageResource *longhorn.EngineImage) ([]*longhorn.Volume, error) {
-	candiates := []*longhorn.Volume{}
+func (ic *EngineImageController) getVolumesForEngineImageUpgrading(volumes map[string]*longhorn.Volume, newEngineImageResource *longhorn.EngineImage) (candidates, inProgress map[string][]*longhorn.Volume) {
+	candidates = make(map[string][]*longhorn.Volume)
+	inProgress = make(map[string][]*longhorn.Volume)
 
 	for _, v := range volumes {
-		// Already attempted to upgrade volume's engine image
-		if v.Spec.EngineImage == newEngineImageResource.Spec.Image {
-			continue
-		}
-		// Volume is in upgrading process. Need to wait for the upgrade to finish.
 		if v.Spec.EngineImage != v.Status.CurrentImage {
+			inProgress[v.Status.OwnerID] = append(inProgress[v.Status.OwnerID], v)
 			continue
 		}
-		if isReady, _ := ic.ds.CheckEngineImageReadinessForVolume(newEngineImageResource.Spec.Image, v.Name, v.Status.CurrentNodeID); !isReady {
-			continue
-		}
-		if !ic.canDoOfflineEngineImageUpgrade(v, newEngineImageResource) && !ic.canDoLiveEngineImageUpgrade(v, newEngineImageResource) {
-			continue
+		canBeUpgraded := ic.canDoOfflineEngineImageUpgrade(v, newEngineImageResource) || ic.canDoLiveEngineImageUpgrade(v, newEngineImageResource)
+		isCurrentEIAvailable, _ := ic.ds.CheckEngineImageReadinessForVolume(v.Status.CurrentImage, v.Name, v.Status.CurrentNodeID)
+		isNewEIAvailable, _ := ic.ds.CheckEngineImageReadinessForVolume(newEngineImageResource.Spec.Image, v.Name, v.Status.CurrentNodeID)
```

Comment:
in the case where the volume is detached, what is the value of status.CurrentNodeID? should these be using the v.Status.OwnerID as well? Or do we only care about the replicas in the case where the volume is detached?

Reply:
Yeah, I think we only care about the replica when volume is detached.